### ISSUE TITLE
Property/attribute in a protocol

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#1368]
How do I specify that a protocol item needs to have a certain property (at least for reading), so that it works with both properties and normal attributes? Neither of these seems to work

a)
```py
class Foo(Protocol):
    x: int

class ShouldBeFoo_1:
    x: int

class ShouldBeFoo_2:
    @property
    def x(self) -> int:
        return 42

def f(foo: Foo) -> int:
    return foo.x ** 2

f(ShouldBeFoo_1())
f(ShouldBeFoo_2())
```
On the last line I get this error:
> Argument of type "ShouldBeFoo_2" cannot be assigned to parameter "foo" of type "Foo" in function "f"
>   "x" is an incompatible type
>     "property" is incompatible with "int"

b)
```py
class Bar(Protocol):
    @property
    def x(self) -> int: ...

class ShouldBeBar_1:
    x: int

class ShouldBeBar_2:
    @property
    def x(self) -> int:
        return 42

def g(bar: Bar) -> int:
    return bar.x ** 2

g(ShouldBeBar_1())
g(ShouldBeBar_2())
```
On the second-to-last line I get:
> Argument of type "ShouldBeBar_1" cannot be assigned to parameter "bar" of type "Bar" in function "g"
>   "x" is an incompatible type
>     "int" is incompatible with "property"

I can fool the typechecker like this:
```py
    if TYPE_CHECKING:
        x: float
        y: float
    else:
        @property
        def x(self) -> float:
            # actual code

        @property
        def y(self) -> float:
            # actual code
```
(and add ` = field(init = False)` in a dataclass), but this obviously isn't a good way

**Comment by decorator-factory** [comment]
I figured that this is related to https://github.com/microsoft/pyright/issues/649, but: I can't just go and "define a property" if I have a dataclass where these `@property`ies are created automatically. More specifically, I have frozen dataclasses, so these function exactly as `@property`.

**Comment by decorator-factory** [comment]
Generally, I think a type with an attribute should satisfy a protocol with a read-only property even if that attribute is writable under the hood. This matches the behaviour from the standard library: `list[int]` satisfies `Sequence[int]`, and `dict[str, int]` satisfies `Mapping[str, int]`, even though they are mutable, because `Sequence` and `Mapping` don't have any promises that the values in them will stay the same.

**Comment by erictraut** [comment]
A property and an attribute do not have the same semantics, so they are not interchangeable. 

This isn't equivalent to `list[int]` being satisfied by `Sequence[int]`. In this case, the callee is promising not to mutate the data in the container. In the case of a protocol that exposes a read-only property, the underlying class is making no promise that the value associated with the property is immutable.

**Comment by decorator-factory** [comment]
So there's no way to  make a protocol that would work for both frozen dataclasses with an `x` attribute and classes with an `x` property?
